### PR TITLE
fix issue with version rollback and thumbnails

### LINKED ISSUE #10559: Thumbnails not regenerated when reverting to an earlier version of a file
### Steps to reproduce
0.  Make sure the Versions app is enabled
1. Create a new document (click on "new"->text file from the "files" view)
2. Enter some basic text; save; close
3. Validate that the thumbnail next to the file is accurate
4.  Edit the file a second time
5.  Validate that the thumbnail changed as expected. 
6.  Highlight the file, select "versions", select previous version
7.  The thumbnail did not revert

I see this with both FF and chrome. I cleared the cache to ensure that this isn't a caching issue.

p1

p2

### Expected behaviour

The thumbnail should revert too

### Actual behaviour

The thumbnail reflects the earlier version

### Server configuration

**ownCloud version:** (see ownCloud admin page)

ownCloud 7.0.1 (daily) Build:2014-08-20T03:18:51+00:00

### PR DESCRIPTION
**Pull request by georgehrke**
fixes #10559
- [x] send hook
- [x] reload preview in js
- [x] add unit test (how to test this?)

### CODE REVIEW

**Comment by georgehrke** [comment]
@DeepDiver1975 Are there any tests for hooks? Not sure how to test this.

**Comment by DeepDiver1975** [comment]
> Are there any tests for hooks? Not sure how to test this.

not really

**Comment by DeepDiver1975** [comment]
is there a bug report which is fixed with this pr? please link - thx

**Comment by georgehrke** [comment]
fixes https://github.com/owncloud/core/issues/10559

**Comment by PVince81** [comment]
@georgehrke maybe you can add a unit test in the versions app to check if the hook is called when calling the rollback function ?

To test a hook, it is possible to register a closure function like this: https://github.com/owncloud/core/blob/master/tests/lib/files/view.php#L1955

File: apps/files_versions/js/versions.js [hunk 1]
```diff
@@ -74,6 +74,20 @@ function revertFile(file, revision) {
...
 					$('#dropdown').closest('tr').find('.modified:first').html(relative_modified_date(revision));
+					//trigger reload of thumbnail
+					var thumbnail = $('#dropdown').closest('tr').find('.thumbnail:first');
+					if (thumbnail.length > 0) {
+						var url = thumbnail.css('background-image');
+						//strip of url(...)
+						url = url.replace('url(','').replace(')','');
+						//add random parameter to force browser to reload
+						url += (url.indexOf('?') === -1) ? '?' : '&';
+						url += 'r=' + Math.round((new Date()).getTime() / 1000);
```
Comment (PVince81) [inline apps/files_versions/js/versions.js]: Use `OC.parseQueryString(url)` to parse a query to a JS hash and `OC.buildQueryString(_.extend(params, {r: ...}))` to convert it back after adding the random param.

Comment (georgehrke) [inline apps/files_versions/js/versions.js]: thx for the hint, will do it this way

**Comment by georgehrke** [comment]
changed the way the random number was added to the url, added a unit test, squashed and rebased.

please review again
thx :)

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
Tested and works with this change :+1:

**Comment by MorrisJobke** [comment]
@Xenopathic @oparoz Do you want to review this?

**Comment by PVince81** [comment]
>is there a bug report which is fixed with this pr? please link - thx

This one: https://github.com/owncloud/core/issues/10559

**Comment by PVince81** [comment]
@georgehrke please rebase.

Note that on master there is no more versions dropdown, so you might want to keep this PR's code for the potential 8.1 backport.

**Comment by georgehrke** [comment]
rebased

File: apps/files_versions/lib/storage.php [hunk 2]
```diff
@@ -315,6 +315,9 @@ public static function rollback($file, $revision) {
 			if (self::copyFileContents($users_view, 'files_versions' . $filename . '.v' . $revision, 'files' . $filename)) {
 				$files_view->touch($file, $revision);
 				Storage::scheduleExpire($uid, $file);
+				\OC_Hook::emit('\OCP\Versions', 'rollback', array(
```
Comment (PVince81) [inline apps/files_versions/lib/storage.php]: I get this wouldn't work with one of the regular write/post-write hooks ?

**Comment by PVince81** [comment]
Tested, works. But am having doubts regarding the hook, see my comment.

File: apps/files_versions/lib/storage.php (hunk 2 above)
Comment (georgehrke) [inline apps/files_versions/lib/storage.php]: Exactly. This was the reason, why the cached preview was not deleted.

**Comment by georgehrke** [comment]
> But am having doubts regarding the hook, see my comment.

https://github.com/owncloud/core/pull/15098#issuecomment-118375493
This comment? 
Could you explain your concerns a bit more? I'm not sure I can follow.

**Comment by PVince81** [comment]
My concern is only about core code having to know about the version hook which is in an app.
If there is no better alternative then I'm fine with this fix :+1:

**Comment by MorrisJobke** [comment]
Tested and still works :+1:

@DeepDiver1975 Merge please

File: apps/files_versions/lib/storage.php (hunk 2 above)
Comment (icewind1991) [inline apps/files_versions/lib/storage.php]: Any idea why the post-write hook didn't work?

**Pull request merged**